# Exploring eBay Car Sales Data

As a prospective buyer of used cars, it is important to understand the factors that can influence their prices. This is especially crucial if you are new to the country and lack awareness about potential considerations. In this project, we will analyze a dataset comprising used cars obtained from _eBay Kleinanzeigen_, the classifieds section of the German eBay website. The primary objective of this project is to perform data cleaning and analysis on the dataset comprising used car listings. Through this process, we aim to ensure the dataset's quality and reliability while extracting meaningful insights from the included information about the used cars. 

The data can be found [here](https://data.world/data-society/used-cars-data). <br>

Some modifications have been made to the original dataset for this project, including:
- A subset of 50,000 data points has been selected from the original dataset for analysis. This subset represents a smaller but still substantial portion of the complete dataset.
- The dataset has been intentionally altered or "dirtied" to replicate characteristics that align more closely with a scraped dataset. This modification aims to create a dataset that is more representative of real-world data with potential inconsistencies and errors.





## Data

First, let's import the data and examine its structure to gain a better understanding of its format and organization.

In [1]:
import pandas as pd
import numpy as np

# Read in .csv file
# encoding popular options: UTF-8(default), Latin-1, Windows-1252
autos = pd.read_csv("autos.csv", encoding="Latin-1") 

In [2]:
# Data table
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
# data information
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head(3)

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   
2   Angebot  $8,990     test   limousine                2009    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

In [5]:
autos.tail(3)

dateCrawled                                name  seller  \
49997  2016-04-02 14:44:48      Fiat_500_C_1.2_Dualogic_Lounge  privat   
49998  2016-03-08 19:25:42  Audi_A3_2.0_TDI_Sportback_Ambition  privat   
49999  2016-03-14 00:42:12                 Opel_Vectra_1.6_16V  privat   

      offerType    price   abtest vehicleType  yearOfRegistration    gearbox  \
49997   Angebot  $13,200     test      cabrio                2014  automatik   
49998   Angebot  $22,900  control       kombi                2013    manuell   
49999   Angebot   $1,250  control   limousine                1996    manuell   

       powerPS   model   odometer  monthOfRegistration fuelType brand  \
49997       69     500    5,000km                   11   benzin  fiat   
49998      150      a3   40,000km                   11   diesel  audi   
49999      101  vectra  150,000km                    1   benzin  opel   

      notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
49997              nein  2016-04-02 00:00:00             0       73430   
49998              nein  2016-03-08 00:00:00             0       35683   
49999              nein  2016-03-13 00:00:00             0       45897   

                  lastSeen  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

In [6]:
autos.describe()

yearOfRegistration       powerPS  monthOfRegistration  nrOfPictures  \
count        50000.000000  50000.000000         50000.000000       50000.0   
mean          2005.073280    116.355920             5.723360           0.0   
std            105.712813    209.216627             3.711984           0.0   
min           1000.000000      0.000000             0.000000           0.0   
25%           1999.000000     70.000000             3.000000           0.0   
50%           2003.000000    105.000000             6.000000           0.0   
75%           2008.000000    150.000000             9.000000           0.0   
max           9999.000000  17700.000000            12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [7]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN


In order to facilitate the data analysis, please take note of the following points:

- `Null` values: There are missing values in five columns with the following indices: `6 (vehicleType)`, `8 (gearbox)`, `10 (model)`, `13 (fuelType)`, and `15 (notRepairedDamage)`. These null values will need to be handled appropriately during the cleaning process.

- Column names: The column names appear to have a combination of upper and lower cases (camelcase) instead of Python's preferred snakecase. It might be beneficial to standardize them by converting all names to either uppercase or lowercase for consistency.

- The `"price"` and the `"odometer"` columns (`column 4 and 11`) requires some formatting adjustments. Specifically, `"price"` needs to be processed to remove any dollar symbols ($) and commas (,), and the comma(,) and km need to be removed from the `"odometer"` so that the values can be analyzed accurately.

- The dataset contains an inconsistency in the "monthOfRegistration" column (column 12), where the minimum value is recorded as 0. This value is invalid since months typically range from 1 to 12. 

- The columns `"seller"`, `"gearbox"`, and `"notRepairedDamage"` are currently in German. It would be beneficial to transform the values in this column to English for better consistency and understanding during the analysis.

- Several columns (e.g. `"seller"` and `"offerType"`) in the dataset have a significant portion of values that are identical.

- The column `"nr_of_pictures"` likely has only one value that can be dropped from the table. 

Taking these factors into account will help ensure that the data is properly prepared for analysis.


## Data Cleaning 
### [1] Rename The Columns
To enhance readability and consistency of the column names, we will convert the column names from camelcase to snakecase. Additionally, we will rephrase certain column names based on the provided data dictionary to offer more descriptive representations.

In [8]:
# Print out column names
cols_copy = autos.columns
print('Original Column Names:\n',cols_copy)

Original Column Names:
 Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [9]:
# Transform camelcase to snakecase
def snakecase(string):
    for char in string:
        if char.isupper():
            string = string.replace(char,"_"+char)
    string = string.lower()
                                    
    return string

In [10]:
# Run through each column name 
new_col = {}
for col in cols_copy:
    new_col[col] = snakecase(col)
    
# Rename the columns
autos.rename(columns=new_col, inplace=True)
print("Updated Column Names:\n",autos.columns)

Updated Column Names:
 Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_p_s', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [11]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price   abtest vehicle_type  year_of_registration    gearbox  \
0    Angebot  $5,000  control          bus                  2004    manuell   
1    Angebot  $8,500  control    limousine                  1997  automatik   
2    Angebot  $8,990     test    limousine                  2009    manuell   

   power_p_s   model   odometer  month_of_registration fuel_type       brand  \
0        158  andere  150,000km                      3       lpg     peugeot   
1        286     7er  150,000km                      6    benzin         bmw   
2        102    golf   70,000km                      7    benzin  volkswagen   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                nein  2016-03-26 00:00:00               0        79588   
1                nein  2016-04-04 00:00:00               0        71034   
2                nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

Following the conversion of the column names to the snakecase format, the dataset exhibits improved readability and clarity.

### [2] Convert Columns `odometer` & `price` to Numeric Type

In [12]:
# strip off comma and km from column odometer
autos["odometer"] = (autos["odometer"].str.replace("km","")
                     .str.replace(",","")
                     .astype(int)
                    )
autos.rename(columns={"odometer":"odometer_km"},inplace=True)

In [13]:
# strip off comma and dollar sign from column price
autos["price"] = (autos["price"].str.replace("$","",regex=True)
                  .str.replace(",","")
                  .astype(int)
                 )

In [14]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   date_crawled           50000 non-null  object
 1   name                   50000 non-null  object
 2   seller                 50000 non-null  object
 3   offer_type             50000 non-null  object
 4   price                  50000 non-null  int32 
 5   abtest                 50000 non-null  object
 6   vehicle_type           44905 non-null  object
 7   year_of_registration   50000 non-null  int64 
 8   gearbox                47320 non-null  object
 9   power_p_s              50000 non-null  int64 
 10  model                  47242 non-null  object
 11  odometer_km            50000 non-null  int32 
 12  month_of_registration  50000 non-null  int64 
 13  fuel_type              45518 non-null  object
 14  brand                  50000 non-null  object
 15  not_repaired_damage

### [3] Remove Outliers in `odometer` & `price` 
In `odometer_km` and `price` columns, we'll check to see if there's unrealistically high or low (outliers) numbers, which we might want to remove.


In [15]:
# Number of unique values in column odometer_km
print("Number of unique values: ",autos["odometer_km"].unique().shape[0])
print(autos["odometer_km"].value_counts().sort_index(ascending=True))

Number of unique values:  13
5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64


In [16]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

The data in `"odometer_km"`  appears to be evenly distributed. As a result, we will retain the original data without making any modifications.

In [17]:
# Number of unique values in column price
print("Number of unique values: ",autos["price"].unique().shape[0])
prices = autos["price"].value_counts()
print(prices.head(50)) 
print(prices.tail(20))
prices_sorted = autos["price"].value_counts().sort_index(ascending=True)
print(prices_sorted.head(50)) 
print(prices_sorted.tail(250))

Number of unique values:  2357
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
1100     376
1300     371
3000     365
550      356
1800     355
5500     340
1250     335
350      335
1600     327
1999     322
400      321
6500     300
1400     292
2800     291
250      291
2300     290
1350     276
1700     268
200      266
450      265
3800     264
3200     261
2900     256
1450     252
4000     246
2999     242
2600     241
1900     239
5000     239
1750     238
Name: price, dtype: int64
369      1
1265     1
21275    1
40400    1
1169     1
13399    1
3120     1
19480    1
23450    1
17899    1
785      1
7840     1
7680     1
14998    1
4755     1
6770     1
61999    1
20987    1
6578     1
48600    1
Name: price, dtype: int64
0      1421
1       156
2         3
3         1

In [18]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Upon analyzing the data, we observe that the mean of the dataset is approximately 10,000. However, it is worth noting that there are outliers with minimum and maximum values of 0 and 100,000,000, respectively. Additionally, there are many zeros and ones which are unrealistic price numbers. Other than the two values, the results obtained from using `value_counts()` shows that the occurrence of values that are excessively small or large is infrequent.

To address this, we will establish a threshold for filtering out these outliers. Specifically, we will exclude values below 50 and above 30,000, thereby ensuring a more reasonable range for analysis.

In [19]:
autos = autos[autos["price"].between(50,30000)]

In [20]:
autos["price"].describe()

count    47567.000000
mean      5198.911199
std       5697.733217
min         50.000000
25%       1200.000000
50%       2999.000000
75%       7000.000000
max      30000.000000
Name: price, dtype: float64


After removing the outliers, the average price now provides a more accurate representation of the data distribution, allowing for a better understanding of the price range within the dataset.

### [4] Data with Date Values

The columns with date information (e.g. `date_crawled`, `date_created`, and `last_seen`) are identified as string. We need to convert these columns into numeric representations to do quantitative analysis. 

|   Column    | Description |
| ----------- | ----------- |
|`date_crawled`|When this ad was first crawled. All field-values are taken from this date.|
|`date_created`|The date on which the eBay listing was created.|
|`last_seen`|When the crawler saw this ad last online.|
|`yearOfRegistration`|The year in which the car was first registered.|

In [21]:
autos.iloc[:,[0,16,19,7]]

date_crawled         date_created            last_seen  \
0      2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1      2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2      2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3      2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4      2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   
...                    ...                  ...                  ...   
49995  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40   
49996  2016-03-28 10:50:25  2016-03-28 00:00:00  2016-04-02 14:18:02   
49997  2016-04-02 14:44:48  2016-04-02 00:00:00  2016-04-04 11:47:27   
49998  2016-03-08 19:25:42  2016-03-08 00:00:00  2016-04-05 16:45:07   
49999  2016-03-14 00:42:12  2016-03-13 00:00:00  2016-04-06 21:18:48   

       year_of_registration  
0                      2004  
1                      1997  
2                      2009  
3                      2007  
4                      2003  
...                     ...  
49995                  2011  
49996                  1996  
49997                  2014  
49998                  2013  
49999                  1996  

[47567 rows x 4 columns]

In [22]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025396
2016-03-06    0.014127
2016-03-07    0.036096
2016-03-08    0.033342
2016-03-09    0.033069
2016-03-10    0.032249
2016-03-11    0.032523
2016-03-12    0.037232
2016-03-13    0.015662
2016-03-14    0.036748
2016-03-15    0.034267
2016-03-16    0.029600
2016-03-17    0.031661
2016-03-18    0.012887
2016-03-19    0.034667
2016-03-20    0.037778
2016-03-21    0.037148
2016-03-22    0.032775
2016-03-23    0.032270
2016-03-24    0.029348
2016-03-25    0.031408
2016-03-26    0.032459
2016-03-27    0.030946
2016-03-28    0.034793
2016-03-29    0.034246
2016-03-30    0.033742
2016-03-31    0.031787
2016-04-01    0.033616
2016-04-02    0.035382
2016-04-03    0.038661
2016-04-04    0.036664
2016-04-05    0.012929
2016-04-06    0.003132
2016-04-07    0.001388
Name: date_crawled, dtype: float64

The data are crawled within a month (2016-03-05 to 2016-04-07).

In [23]:
created = autos['date_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()
print(created.head(10))
print(created[30:40])
print(created.tail(10))

2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000042
2016-01-13    0.000021
2016-01-14    0.000021
Name: date_created, dtype: float64
2016-02-23    0.000084
2016-02-24    0.000042
2016-02-25    0.000063
2016-02-26    0.000042
2016-02-27    0.000126
2016-02-28    0.000210
2016-02-29    0.000168
2016-03-01    0.000105
2016-03-02    0.000105
2016-03-03    0.000883
Name: date_created, dtype: float64
2016-03-29    0.034162
2016-03-30    0.033574
2016-03-31    0.031829
2016-04-01    0.033595
2016-04-02    0.035087
2016-04-03    0.038893
2016-04-04    0.037064
2016-04-05    0.011626
2016-04-06    0.003195
2016-04-07    0.001261
Name: date_created, dtype: float64


Although the listing has been put up since August 2015, most of the listings were put on eBay after March 2016.

In [24]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001093
2016-03-06    0.004373
2016-03-07    0.005445
2016-03-08    0.007463
2016-03-09    0.009734
2016-03-10    0.010743
2016-03-11    0.012530
2016-03-12    0.023987
2016-03-13    0.008977
2016-03-14    0.012719
2016-03-15    0.015998
2016-03-16    0.016461
2016-03-17    0.028318
2016-03-18    0.007316
2016-03-19    0.015935
2016-03-20    0.020687
2016-03-21    0.020687
2016-03-22    0.021507
2016-03-23    0.018710
2016-03-24    0.019867
2016-03-25    0.019320
2016-03-26    0.016797
2016-03-27    0.015683
2016-03-28    0.021044
2016-03-29    0.022537
2016-03-30    0.024954
2016-03-31    0.023945
2016-04-01    0.023041
2016-04-02    0.024975
2016-04-03    0.025333
2016-04-04    0.024765
2016-04-05    0.123951
2016-04-06    0.220405
2016-04-07    0.130700
Name: last_seen, dtype: float64

The time the ads were last viewed have the same duration as the ads that were first detected by the crawling process.

In [25]:
autos["year_of_registration"].describe()

count    47567.000000
mean      2004.567137
std         86.374387
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

Some data might not have the correct year causing the maximum and minimum year to be 9999(into the future) and 1000 (before cars were invented).
Since a car cannot be registered after the listing was viewed, any vehicle with a registration year beyond 2016 is undoubtedly incorrect. However, determining the earliest valid year is a more challenging task. In reality, it could be somewhere within the initial few decades of the 1900s.
<br>
<br>
To assess the situation, let's tally the number of listings that feature cars falling outside the 1900-2016 timeframe. Based on this count, we can determine whether it is advisable to completely remove those rows or if we need to employ additional customized logic.

In [26]:
# Decide the highest and lowest acceptable values for the `year_of_registration` column
years = autos['year_of_registration'].value_counts(normalize=True, dropna=False).sort_index()
# print(years)
print(years.head(20))
print(years[40:50])
print(years.tail(20))

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000042
1910    0.000042
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000084
1938    0.000021
1939    0.000021
1941    0.000042
1948    0.000021
1950    0.000021
1951    0.000021
1952    0.000021
1953    0.000021
1954    0.000042
1956    0.000084
Name: year_of_registration, dtype: float64
1977    0.000420
1978    0.000862
1979    0.000694
1980    0.001661
1981    0.000568
1982    0.000841
1983    0.001051
1984    0.001072
1985    0.001955
1986    0.001472
Name: year_of_registration, dtype: float64
2009    0.043370
2010    0.032459
2011    0.032459
2012    0.025038
2013    0.014863
2014    0.011605
2015    0.005613
2016    0.024954
2017    0.029054
2018    0.009776
2019    0.000042
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000063
5911    0.000021
8888    0.000021
9000    0.000021
9999    0.000063
Name: year_of_registration, dtype: float64


Considering the significant time gap between 1800 and 1910, and the relatively higher accuracy of records from 1910 onwards, we will establish the lower threshold as 1900. This decision is made to enhance the credibility of our data. Furthermore, we will set the upper bound at 2016 to maintain a realistic timeframe.

In [27]:
# Keep cars that have registration year between 1900 and 2016
autos = autos[autos["year_of_registration"].between(1900,2016)]
autos["year_of_registration"].describe()

count    45700.000000
mean      2002.836652
std          7.002236
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       2016.000000
Name: year_of_registration, dtype: float64

## Data Analysis
### Brand vs. Price
When dealing with car data, it is common to examine variations among different car brands. Utilizing aggregation can provide insights into the `brand` column, helping us gain a better understanding of the data.

In [28]:
# Identify the values we want to aggregate by
# Brand - Percentage
autos['brand'].value_counts(normalize=True)

volkswagen        0.213370
opel              0.109081
bmw               0.108884
mercedes_benz     0.094989
audi              0.084376
ford              0.070832
renault           0.047965
peugeot           0.030416
fiat              0.026149
seat              0.018534
skoda             0.016674
nissan            0.015536
mazda             0.015405
smart             0.014420
citroen           0.014267
toyota            0.012932
hyundai           0.010219
volvo             0.009256
sonstige_autos    0.009015
mini              0.008884
mitsubishi        0.008271
honda             0.007943
kia               0.007221
alfa_romeo        0.006740
suzuki            0.006039
chevrolet         0.005711
chrysler          0.003589
porsche           0.002998
dacia             0.002691
daihatsu          0.002560
jeep              0.002123
subaru            0.002123
saab              0.001685
land_rover        0.001663
daewoo            0.001510
jaguar            0.001510
trabant           0.001400
r

For analysis purposes, we will focus on car brands that account for more than 5% of the total listings.

In [29]:
# Selected brands
selected_brand = autos['brand'].value_counts(normalize=True)[0:6]
print(selected_brand)

volkswagen       0.213370
opel             0.109081
bmw              0.108884
mercedes_benz    0.094989
audi             0.084376
ford             0.070832
Name: brand, dtype: float64


In [30]:
brand_mean_price = {}
for idx in range(0,selected_brand.size):
    brand = selected_brand.index[idx]
    rows = autos['brand'] == brand
    brand_price = autos.loc[rows,['brand','price']]
    mean_price = brand_price['price'].sum()/brand_price['price'].size
    brand_mean_price[brand] = mean_price
print(brand_mean_price)

{'volkswagen': 5194.59850271767, 'opel': 2983.0866599799397, 'bmw': 7450.8563102893895, 'mercedes_benz': 7467.679797281732, 'audi': 7985.665197095435, 'ford': 3543.1451961692924}


Opel and Ford have comparatively lower prices when compared to other brands. Collectively, they constitute approximately 20% of the total listings. On the other hand, BMW, Mercedes Benz, and Audi exhibit higher average prices. Volkswagen, with more than double the number of listings compared to the other brands, falls in between the lower and higher price ranges.
<br>
### Brand, Price, & Mileage
To gain insights into the top 6 brands, we will employ aggregation techniques to analyze the average mileage of their cars. Additionally, we will explore any potential correlations between the average mileage and the mean price of these brands.
<br>
Let's start with calculating the mean mileage for the top 6 brands.

In [31]:
# Calculate the mean mileage
brand_mean_mile = {}
for idx in range(0,selected_brand.size):
    brand = selected_brand.index[idx]
    rows = autos['brand'] == brand
    brand_mile = autos.loc[rows,['brand','odometer_km']]
    mean_mile = brand_mile['odometer_km'].sum()/brand_mile['odometer_km'].size
    brand_mean_mile[brand] = mean_mile
print(brand_mean_mile)

{'volkswagen': 129312.37821761871, 'opel': 129466.39919759278, 'bmw': 134607.11414790998, 'mercedes_benz': 133021.1932734393, 'audi': 132653.0082987552, 'ford': 124700.33982082175}


In [32]:
# Convert both dictionaries to series objects using series constructor
bmp_series = pd.Series(brand_mean_price)
bmm_series = pd.Series(brand_mean_mile)

top6brands = pd.DataFrame(bmp_series, columns=['mean_price'])
top6brands['mean_mile'] = bmm_series

In [33]:
top6brands

mean_price      mean_mile
volkswagen     5194.598503  129312.378218
opel           2983.086660  129466.399198
bmw            7450.856310  134607.114148
mercedes_benz  7467.679797  133021.193273
audi           7985.665197  132653.008299
ford           3543.145196  124700.339821

Based on the provided table, it appears that the average mileages among different brands are relatively similar, and there seems to be limited correlation between the average mileage and the mean price across the various brands.

### Price, Mileage & Year
When considering a used car, mileage and the year of manufacture are two crucial factors that significantly impact a buyer's decision. In this analysis, we will examine the relationship between price and varying mileage and year of the vehicles.

In [34]:
# Show the information of price to have a glimpse into the price range
autos["price"].describe()

count    45700.000000
mean      5266.661838
std       5746.884425
min         50.000000
25%       1249.750000
50%       3000.000000
75%       7250.000000
max      30000.000000
Name: price, dtype: float64

We will first separate the price into four different groups: 
1. \\$50 - \\$1,000
2. \\$1,000 - \\$10,000
3. \\$10,000 - \\$20,000
4. \\$20,000 - \\$30,000

In [35]:
bool_1 = autos["price"].between(50,1000)
bool_2 = autos["price"].between(1001,10000)
bool_3 = autos["price"].between(10001,20000)
bool_4 = autos["price"].between(20001,30000)

print("Average Year:{:.2f} Odometer(km):{:.2f} for $50-1,000".format(autos.loc[bool_1,"year_of_registration"].mean(),autos.loc[bool_1,"odometer_km"].mean()))
print("Average Year:{:.2f} Odometer(km):{:.2f} for $1,000-10,000".format(autos.loc[bool_2,"year_of_registration"].mean(),autos.loc[bool_2,"odometer_km"].mean()))
print("Average Year:{:.2f} Odometer(km):{:.2f} for $10,000-20,000".format(autos.loc[bool_3,"year_of_registration"].mean(),autos.loc[bool_3,"odometer_km"].mean()))
print("Average Year:{:.2f} Odometer(km):{:.2f} for $20,000-30,000".format(autos.loc[bool_4,"year_of_registration"].mean(),autos.loc[bool_4,"odometer_km"].mean()))

Average Year:1998.80 Odometer(km):138382.31 for $50-1,000
Average Year:2002.90 Odometer(km):130545.41 for $1,000-10,000
Average Year:2007.85 Odometer(km):101286.58 for $10,000-20,000
Average Year:2009.29 Odometer(km):75421.53 for $20,000-30,000


As anticipated, the price of used cars generally rises as the vehicles become newer and have lower mileage. The provided figures offer insights on selecting a car based on your budget, providing an idea of what to consider when making a purchase decision.

## Conclusions

Based on our analysis of data from _eBay Kleinanzeigen_, the following findings emerged:
- The brands with the highest number of listings are Volkswagen, Opel, BMW, Mercedes Benz, Audi, and Ford.
- Opel and Ford tend to have relatively lower prices compared to other brands.
- BMW, Mercedes Benz, and Audi exhibit higher average prices.
- Volkswagen has the highest number of listings among all brands, and its prices fall within a range between the lower and higher ends.
- The difference in mileage does not appear to be the primary factor contributing to price variations across different brands.
- We have provided average mileage and registered year for different price ranges as a useful reference for prospective car buyers.
